### PR TITLE
[Pipelined]Patch aioeventlet python crash

### PR DESCRIPTION
**Pull request by koolzz**
Signed-off-by: Nick Yurchenko <koolzz@fb.com>

## Summary

Attempts to gracefully recover from
```
Apr 23 13:42:06 phy-u1 pipelined[3348494]: Traceback (most recent call last):
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     return _run_code(code, main_globals, None,
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     exec(code, run_globals)
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/main.py", line 190, in <module>
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     main()
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/magma/pipelined/main.py", line 174, in main
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     service.run()
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/magma/common/service.py", line 243, in run
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     self._loop.run_forever()
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/aioeventlet.py", line 259, in run_forever
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     super(EventLoop, self).run_forever()
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/lib/python3.8/asyncio/base_events.py", line 570, in run_forever
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     self._run_once()
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/lib/python3.8/asyncio/base_events.py", line 1823, in _run_once
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     event_list = self._selector.select(timeout)
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/aioeventlet.py", line 173, in select
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     events = self._read_events()
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/aioeventlet.py", line 168, in _read_events
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     key = self.get_key(fd)
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/lib/python3.8/selectors.py", line 192, in get_key
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     raise KeyError("{!r} is not registered".format(fileobj)) from None
Apr 23 13:42:06 phy-u1 pipelined[3348494]: KeyError: '20 is not registered'
Apr 23 13:42:06 phy-u1 pipelined[3348494]: Exception in thread Thread-1:
Apr 23 13:42:06 phy-u1 pipelined[3348494]: Traceback (most recent call last):
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     self.run()
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/sentry_sdk/integrations/threading.py", line 69, in run
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     reraise(*_capture_exception())
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/sentry_sdk/_compat.py", line 54, in reraise
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     raise value
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/sentry_sdk/integrations/threading.py", line 67, in run
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     return old_run_func(self, *a, **kw)
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/lib/python3.8/threading.py", line 870, in run
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     self._target(*self._args, **self._kwargs)
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/grpc/_server.py", line 875, in _serve
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     if not _process_event_and_continue(state, event):
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/grpc/_server.py", line 839, in _process_event_and_continue
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     rpc_state, rpc_future = _handle_call(event, state.generic_handlers,
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/grpc/_server.py", line 749, in _handle_call
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     return _handle_with_method_handler(rpc_event, method_handler,
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/grpc/_server.py", line 725, in _handle_with_method_handler
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     return state, _handle_unary_unary(rpc_event, state,
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/local/lib/python3.8/dist-packages/grpc/_server.py", line 629, in _handle_unary_unary
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     return thread_pool.submit(_unary_response_in_pool, rpc_event, state,
Apr 23 13:42:06 phy-u1 pipelined[3348494]:   File "/usr/lib/python3.8/concurrent/futures/thread.py", line 181, in submit
Apr 23 13:42:06 phy-u1 pipelined[3348494]:     raise RuntimeError('cannot schedule new futures after '
Apr 23 13:42:06 phy-u1 pipelined[3348494]: RuntimeError: cannot schedule new futures after interpreter shutdown
```

## Test Plan
@ardzoht , @pshelar  is there an easy way to reproduce?

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/deploy/roles/magma/files/patches/aioeventlet_fd_exception.patch [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
++++ b/aioeventlet.py
+@@ -159,7 +159,10 @@ class _Selector(asyncio.selectors._BaseSelectorImpl):
+         self._notified = {}
+         ready = []
+         for fd, events in notified.items():
+-            key = self.get_key(fd)
++            try:
++                key = self.get_key(fd)
++            except KeyError:
++                continue
```
Comment (koolzz) [inline lte/gateway/deploy/roles/magma/files/patches/aioeventlet_fd_exception.patch:11]: not quite sure if I should be unregistering the FD here... Will add a log here and hopefully reproduce somehow

**Comment by ardzoht** [comment]
@koolzz this is easily reproducible on spirent, once this merges let's backport it and run a test with this cc @pshelar

**Comment by koolzz** [comment]
> @koolzz this is easily reproducible on spirent, once this merges let's backport it and run a test with this cc @pshelar

I can just patch it on spirient it's a tiny change, will test out tmrw

**Review by ardzoht — APPROVED** [review]
lgtm

File: lte/gateway/deploy/roles/magma/tasks/main.yml [hunk 2]
```diff
@@ -405,3 +405,15 @@
...
+  patch:
+    src: patches/aioeventlet_fd_exception.patch
+    dest: /usr/local/lib/python3.8/dist-packages/aioeventlet.py
+  when: ansible_distribution == "Ubuntu"
+
+- name: Patch python aioeventlet lib crash debian
+  patch:
+    src: patches/aioeventlet_fd_exception.patch
+    dest: /usr/lib/python3/dist-packages/aioeventlet.py
+  when: ansible_distribution == "Debian"
```
Comment (koolzz) [inline lte/gateway/deploy/roles/magma/tasks/main.yml:419]: @tmdzk what magic is needed to apply this for release?

Comment (pshelar) [reply lte/gateway/deploy/roles/magma/tasks/main.yml:419]: We need to add symb-link in patch dir for https://github.com/magma/magma/blob/master/third_party/build/bin/aioeventlet_build.sh.
also update iteration for package.

Comment (koolzz) [reply lte/gateway/deploy/roles/magma/tasks/main.yml:419]: I'm a bit confused is there another patch dir other than lte/gateway/deploy/roles/magma/files/patches/ ? 
Do I just need to bump ITERATION to 2, does it build automatically in CI or?

Comment (pshelar) [reply lte/gateway/deploy/roles/magma/tasks/main.yml:419]: you need to create symbolic link in this dir: https://github.com/magma/magma/tree/master/third_party/build/patches/aioeventlet

**Review by pshelar — APPROVED** [review]

File: lte/gateway/deploy/roles/magma/tasks/main.yml (hunk 2 above)
Comment (pshelar) [reply lte/gateway/deploy/roles/magma/tasks/main.yml:419]: its manual build and upload to CI.

**Pull request merged**